src/ui/default_viewer_setup.ts

```diff
@@ -14,26 +14,124 @@
-import type { MinimalViewerOptions } from "#src/ui/minimal_viewer.js";
 import { bindTitle } from "#src/ui/title.js";
+import type { Tool } from "#src/ui/tool.js";
+import { restoreTool } from "#src/ui/tool.js";
 import { UrlHashBinding } from "#src/ui/url_hash_binding.js";
+import {
+  verifyObject,
+  verifyObjectProperty,
+  verifyString,
+} from "#src/util/json.js";
 
 declare let NEUROGLANCER_DEFAULT_STATE_FRAGMENT: string | undefined;
 
+type CustomToolBinding = {
+  layer: string;
+  tool: unknown;
+  provider?: string;
+};
+
+type CustomBindings = {
+  [key: string]: CustomToolBinding | string | boolean;
+};
+
+declare const CUSTOM_BINDINGS: CustomBindings | undefined;
+export const hasCustomBindings =
+  typeof CUSTOM_BINDINGS !== "undefined" &&
+  Object.keys(CUSTOM_BINDINGS).length > 0;
+
 /**
  * Sets up the default neuroglancer viewer.
  */
-export function setupDefaultViewer(options?: Partial<MinimalViewerOptions>) {
-  const viewer = ((<any>window).viewer = makeDefaultViewer(options));
+export function setupDefaultViewer() {
+  const viewer = ((<any>window).viewer = makeDefaultViewer());
   setDefaultInputEventBindings(viewer.inputEventBindings);
 
+  const bindNonLayerSpecificTool = (
+    obj: unknown,
+    toolKey: string,
+    desiredLayerType: UserLayerConstructor,
+    desiredProvider?: string,
+  ) => {
+    let previousTool: Tool<object> | undefined;
+    let previousLayer: UserLayer | undefined;
+    if (typeof obj === "string") {
+      obj = { type: obj };
+    }
+    verifyObject(obj);
+    const type = verifyObjectProperty(obj, "type", verifyString);
+    viewer.bindAction(`tool-${type}`, () => {
+      const acceptableLayers = viewer.layerManager.managedLayers.filter(
+        (managedLayer) => {
+          const correctLayerType =
+            managedLayer.layer instanceof desiredLayerType;
+          if (desiredProvider && correctLayerType) {
+            for (const dataSource of managedLayer.layer?.dataSources || []) {
+              const protocol = viewer.dataSourceProvider.getProvider(
+                dataSource.spec.url,
+              )[2];
+              if (protocol === desiredProvider) {
+                return true;
+              }
+            }
+            return false;
+          } else {
+            return correctLayerType;
+          }
+        },
+      );
+      if (acceptableLayers.length > 0) {
+        const firstLayer = acceptableLayers[0].layer;
+        if (firstLayer) {
+          if (firstLayer !== previousLayer) {
+            previousTool = restoreTool(firstLayer, obj);
+            previousLayer = firstLayer;
+          }
+          if (previousTool) {
+            viewer.activateTool(toolKey, previousTool);
+          }
+        }
+      }
+    });
+  };
+
+  if (hasCustomBindings) {
+    for (const [key, val] of Object.entries(CUSTOM_BINDINGS!)) {
+      if (typeof val === "string") {
+        viewer.inputEventBindings.global.set(key, val);
```

Comment:
Note that in addition to the global map there are several others that are relevant in default_input_event_bindings.ts --- in particular the default slice view and perspective view bindings.  Perhaps this should account for that?

Reply:
I added a "context" property, one of `"global" | "perspectiveView" | "sliceView"`. I could also call it map or actionMap. Restructured the json object to support it.


```json
{
  "keym": {
    "action": {
      "layerType": "segmentation",
      "tool": "grapheneMergeSegments",
      "provider": "graphene"
    }
  },
  "keyx": {
    "action": false
  },
  "control+shift+keyx": {
    "action": "clear-segments",
    "context": "sliceView"
  }
}
```